version-negotiation.md

```diff
@@ -0,0 +1,74 @@
+# Explainer: version negotiation for origin policies
+
+This explainer lays out the problem and solution space for how origin policies can be updated, in a privacy-preserving way.
+
+## Use cases and requirements
+
+### Async update ([#10](https://github.com/WICG/origin-policy/issues/10))
+
+Web applications want to be able to decide that a previously-downloaded origin policy, or even no policy at all, is "good enough", and that they don't need to block the page load waiting for the newest origin policy version.
+
+### Sync update
+
+Web applications want to be able to decide that a certain policy is mandatory, and the page cannot load until it arrives and is applied.
+
+Similarly, web applications may have previously sent a bad policy, which they need to replace before any further page loads continue.
+
+### Kill switch
+
+There needs to be a way for a page to remove any previous origin policies, before the page loads. This is essentially a special case of sync update, which doesn't require loading a new policy since the browser already knows what the null policy looks like.
+
+### Privacy-preserving
+
+Any scenario in which the browser communicates previously-sent information back to the server or website can be (ab)used as a cookie. Previous incarnations of origin policy communicated the currently-downloaded policy version. But even the behaviors controlled by origin policy could be used in this manner.
+
+## Proposed model
+
+Conceptually, origin policies are stored in the HTTP cache, under the URL `$origin/.well-known/origin`. In particular, this means they are double-keyed [in the same way as the HTTP cache](https://github.com/WICG/origin-policy/issues/23): that is, by (origin of document, origin of top-document). This prevents them from being used for cross-site tracking.
```

Comment:
I don't get this bit. Is the "Conceptionally [...] under the URL [...]" meant 
- as change in the policy store, where now multiple policy versions are found in the same resource/JSON doc?
- or is this just meant as a specification tool, that whatever policies we retrieve should be stored as-if they were all under the same resrouce?

Reply:
My intent is that there is only a single policy per HTTP cache key. It is conceptually stored in the HTTP cache, and should behave in that way (e.g. HTTP cache expiration semantics, clearing the cache clears the policy, etc.). Implementations might implement something different under the hood, but the behavior should be the same as any other document-in-the-cache.

Does that work? It seems like it might not, if per @annevk's comment in https://github.com/WICG/origin-policy/issues/10#issuecomment-546235453 we cannot accept having a single policy per HTTP cache key, and need to be able to juggle multiple at once in order to deal with async updates.

Reply:
My comment does not constrain that as you probably want to copy the policy anyway when applying it to a document or agent cluster at which point you can change the backing data.

Reply:
> My intent is that there is only a single policy per HTTP cache key. It is conceptually stored in the HTTP cache, and should behave in that way

In particular, the spec would be similar to the processing model outlined below; the step "Checks if the has something cached, and not expired, for `$that_origin/.well-known/origin-policy`" means literally checking the HTTP cache, at least in spec terms. And that cache would be updated by any subsequent fetches against that URL.

Reply:
I see. That does offer an interesting foot-gun, if you have a typo in the policy name in the policy doc (or forgot to change it from a previous version), then you have bricked your site. Potentially, a similar thing could happen if in your infrastructure changes to resource content is updated at different speeds from resource headers, and if you aren't careful with the allowed version ranges.

(Honestly, I'm unsure how much of an argument this is. Because a website operator certainly has all the tools to deploy things rightly. In an unrelated thing I'm working on, a substantial amount of early feedback was around such deployability issues, so I'm trying a bit harder to anticipate such feedback this time around. I guess it might make sense to, instead of overthinking it, try to get early feedback from potential adopters.)